packages/messaging/messaging.ts

```diff
@@ -0,0 +1,171 @@
+import type { Messages, PayloadAndResponse } from './type';
+
+type MessageType = keyof Messages extends never ? string : keyof Messages;
+// @ts-expect-error This is a placeholder type
+type Message<K extends MessageType = unknown> = PayloadAndResponse & Messages[K] & { type: K };
+
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+type Payload<Type extends MessageType> = Messages[Type] extends PayloadAndResponse<infer P, unknown> ? P : never;
+
+export type Response<Type extends MessageType> =
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  Messages[Type] extends PayloadAndResponse<unknown, infer D> ? D : never;
+
+const ERROR_SUFFIX = '__Error';
+
+function sendByPort<T extends MessageType>(type: T, payload?: Payload<T>): Promise<Response<T>> {
+  return new Promise((resolve, reject) => {
+    const port = chrome.runtime.connect();
+
+    port.onMessage.addListener((message: Message<T>) => {
+      const type = message.type;
+      const response = message.response as Response<T>;
```

Comment:
```suggestion
    port.onMessage.addListener(({ type, response }: Message<T>) => {
```
And type for response should be taken from Message